projects/2-state-machines/chern.ipynb

```diff
@@ -0,0 +1,492 @@
+   "source": [
+    "# FIXME: if and only if `sm` accepts `string`.\n",
+    "def check(sm, string):\n",
+    "    for x in string:\n",
+    "        if x not in sm.out_edges():\n",
+    "            return False\n",
+    "        sm.advance(x)\n",
+    "    return sm.is_terminal()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It should give the correct results for at least these examples, and we encourage you to try out more tests."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 34,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "# Test suite 1. Should not throw any errors.\n",
+    "# (First, we define a function to produce the test state machine so we can use a fresh one every time.)\n",
+    "def test1():\n",
+    "    # This one accepts \"h(e(llo|y) | i(hi)*)\n",
+    "    return StateMachine(\n",
+    "        [(\"s1\", \"h\", \"s2\"), (\"s2\", \"e\", \"s3\"), (\"s3\", \"y\", \"s4\"), \n",
+    "         (\"s3\", \"l\", \"s5\"), (\"s5\", \"l\", \"s6\"), (\"s6\", \"o\", \"s7\"),\n",
+    "         (\"s2\", \"i\", \"s8\"), (\"s8\", \"h\", \"s9\"), (\"s9\", \"i\", \"s8\")\n",
+    "        ],\n",
+    "        \"s1\",\n",
+    "        [\"s4\", \"s7\", \"s8\"]\n",
+    "    )\n",
+    "\n",
+    "assert check(test1(), \"hey\")\n",
+    "assert check(test1(), \"hi\")\n",
+    "assert check(test1(), \"hello\")\n",
+    "assert check(test1(), \"hihi\")\n",
+    "assert check(test1(), \"hihihi\")\n",
+    "assert not check(test1(), \"greetings\")\n",
+    "assert not check(test1(), \"hel\")\n",
+    "assert not check(test1(), \"helloy\")\n",
+    "assert not check(test1(), \"hih\")\n",
+    "assert not check(test1(), \"heyhey\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, define a state machine which accepts email addresses consisting only of the letter \"x\" and the symbols \"@\" and \".\" (in other words, you don't need a ton of edges). Devise tests for it below (try to be sinister---get a partner to come up with adversarial examples too)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 53,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def test2():\n",
+    "    # TODO: define transition system here:\n",
+    "    return StateMachine([(\"s1\", \"x\", \"s2\"),\n",
+    "                        (\"s2\", \"@\", \"s3\"), (\"s2\", \"x\", \"s11\"),\n",
+    "                        (\"s3\", \"x\", \"s4\"), (\"s4\", \"x\", \"s5\"), (\"s4\", \".\", \"s5\"), (\"s5\", \"x\", \"s6\"),\n",
+    "                        (\"s5\", \".\", \"s6\"), (\"s6\", \"x\", \"s7\"),\n",
+    "                        (\"s11\", \"x\", \"s12\"), (\"s11\", \"@\", \"s3\"), (\"s12\", \"x\", \"s2\"), (\"s12\", \"@\", \"s3\")],\n",
+    "                        \"s1\",\n",
+    "                        [\"s6\", \"s7\"])\n",
+    "\n",
+    "# TODO: tests go here\n",
```

Comment:
One of my email addresses is `jcosborn@soe.ucsc.edu`, and it looks like your regex won't take `x@x.x.x`. Also, your state machine is a little more complicated than it needs to be: why do `s11` and `s12` exist? It seems like the state where we're waiting for either an `@` or a `x` can just be `s2`, and it can self-loop on seeing an `x`. Similarly, `s4` can self-loop on seeing more letters. I was able to do it with 7 states, of which 1 was accepting. Might be good to try for that!

I found it easiest to write the string recognizer as a regular expression, then manually build a state machine that realized it.